Gio/tui

sinkingsugar opened the pull request:
This commit reformats the warmup method in the Separator structure to enhance code readability by adding line breaks. Additionally, it removes redundant element initialization in the activate method, streamlining the element management process within the TUI module.


<!--
Before submitting your PR, please review the following checklist:

- [ ] **DO NOT** submit a PR without having discussed the change first in a related issue. If none exist, create one first.
- [ ] **CONSIDER** adding a unit test if your PR resolves an issue.
- [ ] **DO** keep pull requests small so they can be easily reviewed.
- [ ] **DO** make sure all unit tests pass.
- [ ] **DO** make sure not to introduce any new compiler warnings.
- [ ] **AVOID** breaking the continuous integration build.
- [ ] **AVOID** making significant changes to the overall architecture.
-->

sinkingsugar committed:
Add FTXUI submodule and integrate TUI module

This commit adds the FTXUI submodule to the project and includes a new TUI module with its implementation in C++. The TUI module provides components for creating and rendering terminal user interfaces, enhancing the overall functionality of the application. Additionally, the CMake configuration has been updated to include the new module and its dependencies.

sinkingsugar committed:
Enhance TUI module by adding inner elements management

This commit introduces a new member variable `_innerElements` to the TUI module's HBox structure and updates the `activate` method to clear and prepare these inner elements for context. This change improves the management of inner elements within the terminal user interface.

sinkingsugar committed:
Update TUI module to utilize inner elements in HBox activation

This commit modifies the `activate` method in the HBox structure to use the `_innerElements` member variable when creating the hbox element. This change enhances the functionality of the TUI module by ensuring that the inner elements are properly incorporated into the terminal user interface.

sinkingsugar committed:
Add TUI.VBox and TUIText components to TUI module

This commit introduces the VBox and TUIText structures to the TUI module, allowing for the creation of vertical boxes and text elements within the terminal user interface. The TUI module's registration function is updated to include these new components, enhancing the overall functionality and flexibility of the TUI system. Additionally, the test script is modified to utilize the new TUI.VBox and TUI.Text components for improved demonstration and logging.

sinkingsugar committed:
Editing file tui.cpp

sinkingsugar committed:
Enhance TUI module with improved element management and logging

This commit adds a destructor to the TUIElement structure for logging when elements are destroyed. It also refactors the warmup and cleanup methods in various structures to use consistent parameter names. Additionally, the activate method in the box shard macro now logs the size of inner elements, improving debugging and traceability within the TUI module.

sinkingsugar committed:
Editing file tui.cpp

sinkingsugar committed:
Add Separator component to TUI module and update TUIText activation

This commit introduces a new Separator structure to the TUI module, allowing for the addition of separator elements within the terminal user interface. The TUIText activation method is also updated to ensure proper management of inner elements, enhancing the overall functionality of the TUI system. Additionally, the test script is modified to demonstrate the use of the new Separator component within a horizontal box layout.

sinkingsugar committed:
Refactor warmup method in Separator structure for improved readability

This commit reformats the warmup method in the Separator structure to enhance code readability by adding line breaks. Additionally, it removes redundant element initialization in the activate method, streamlining the element management process within the TUI module.

sinkingsugar committed:
Add Tick component to TUI module for rendering elements

This commit introduces a new Tick structure to the TUI module, enabling the rendering of TUI elements into strings. The Tick component includes methods for warmup, cleanup, and activation, enhancing the functionality of the TUI system. Additionally, the test script is updated to demonstrate the use of the new Tick component in the terminal user interface.

sinkingsugar committed:
Enhance Tick activation method in TUI module with quit handling

This commit updates the activate method in the Tick structure to include a check for application quit status, allowing for graceful termination of the flow. The method now logs a debug message when the application is quitting and uses RunOnce instead of RunOnceBlocking, improving responsiveness in the TUI module.

sinkingsugar committed:
wip button

sinkingsugar committed:
Update help text for Button and Tick structures in TUI module for clarity

This commit modifies the help strings for the Button and Tick structures to improve clarity. The Button help text now includes a period at the end, while the Tick help text is updated to better describe its functionality in handling interactive rendering and event loop management.

sinkingsugar committed:
Refactor Button and Tick structures in TUI module for improved element management

This commit updates the Button structure to utilize an optional component for better handling of button rendering. The Tick structure's warmup method is refactored to use a dedicated getElement method, enhancing code clarity and maintainability. Additionally, minor formatting adjustments are made in the test script for consistency.

sinkingsugar committed:
Enhance Button structure in TUI module with animated option for improved user interaction

This commit updates the Button structure to include an animated option for button rendering, enhancing the visual feedback during user interactions. The activation method remains unchanged in functionality but is now more visually appealing, contributing to a better user experience in the TUI module.

sinkingsugar committed:
Update TUI module to change screen mode from Fullscreen to TerminalOutput

This commit modifies the Tick structure to set the screen mode to TerminalOutput instead of Fullscreen, enhancing compatibility with various terminal environments. This change aims to improve the user experience by ensuring better rendering across different terminal setups.

sinkingsugar committed:
Enhance TUI module by adding component management to TUIElement and TUIInnerElements

This commit updates the TUIElement and TUIInnerElements structures to include a components field, allowing for better management of UI components. The reset and clear methods are modified to handle the new components field, ensuring proper cleanup. Additionally, the Tick structure is updated to integrate the root component, improving the rendering process of TUI elements.

sinkingsugar committed:
Update TUI test script to include logging for button interactions

This commit modifies the test script for the TUI module by adding logging functionality to the button interactions. The changes ensure that both the "Hello" and "Ciao" button actions log their outputs, enhancing the debugging and tracking of user interactions within the terminal user interface.

guusw committed:
Automatically disable terminal output

sinkingsugar committed:
Merge branch 'gio/fast-string-review' into gio/tui

sinkingsugar committed:
Enhance TUI rendering by forcing a screen redraw on each tick

This commit adds a call to _screen.RequestAnimationFrame() in the TUI tick function to ensure the screen is redrawn during each loop iteration. Additionally, new test cases have been added to validate the rendering behavior with memoization and random text generation.

sinkingsugar committed:
Implement Track shard for variable change tracking and enhance TUI components

This commit introduces a new Track shard that monitors specified variables and triggers an action when changes occur. Additionally, TUI components have been updated to utilize ftxui::Component instead of ftxui::Element, improving the rendering process. The TUI test script has been modified to demonstrate the new tracking functionality and ensure proper integration with existing components.

sinkingsugar committed:
Merge remote-tracking branch 'origin/devel' into gio/tui

sinkingsugar committed:
Enhance TUI.Input component with placeholder and password support

This commit introduces a constructor for the TUIInput struct, adding support for placeholder text and password input functionality. The input handling logic has been updated to utilize these new parameters, ensuring that the component behaves correctly based on user input. Additionally, the TUI test script has been modified to include a test case for the new input functionality.

sinkingsugar committed:
Enhance TUI.Input with multiline support and OnEnter action

This commit adds a new multiline parameter to the TUIInput struct, allowing for multiline input functionality. Additionally, an OnEnter action is implemented, enabling users to define behavior when the input is submitted. The TUI test script is updated to demonstrate the new OnEnter feature, ensuring proper integration and functionality.

sinkingsugar committed:
Refactor TUIInput OnEnter action to clean buffer and update value

This commit enhances the OnEnter action in the TUIInput struct by adding logic to remove trailing newlines from the input buffer before updating the value variable. The buffer is cleared after processing the input, ensuring a clean state for subsequent entries. This improves input handling and user experience in the TUI component.

sinkingsugar committed:
Merge remote-tracking branch 'origin/devel' into gio/tui

sinkingsugar committed:
Enhance TUIText component with border and flex parameters

This commit introduces two new parameters, Border and Flex, to the TUIText struct, allowing users to customize the text element's appearance in the TUI context. The rendering logic is updated to conditionally apply a border and flex behavior based on these parameters. Additionally, the TUI test script is modified to demonstrate the new functionality, ensuring proper integration and usage.

sinkingsugar committed:
Refactor TUIInput OnEnter action to conditionally execute based on presence of _onEnter

This commit modifies the TUIInput struct to ensure the OnEnter action is only set if the _onEnter variable is defined. The logic for cleaning the input buffer and updating the value remains intact, enhancing the robustness of input handling in the TUI component.

sinkingsugar committed:
Enhance TUIText component with AlignRight parameter

This commit introduces a new AlignRight parameter to the TUIText struct, allowing users to align text to the right within the TUI context. The rendering logic is updated to conditionally apply right alignment based on this parameter. Additionally, the TUI test script is modified to demonstrate the new functionality, ensuring proper integration and usage.

sinkingsugar committed:
Enhance TUIText component with color and background color parameters

This commit introduces two new parameters, Color and BackgroundColor, to the TUIText struct, allowing users to customize the text and background colors in the TUI context. The rendering logic is updated to apply these colors accordingly. Additionally, the TUI test script is modified to demonstrate the new functionality, ensuring proper integration and usage.

sinkingsugar committed:
Update memoize logic to conditionally activate tracking based on _shouldActivate flag; enhance TUI test script with additional separators for improved layout.

sinkingsugar committed:
Add TUISplit component for multi-pane layout in TUI

This commit introduces the TUISplit struct, which allows users to create a multi-pane layout by surrounding an input element with customizable top, bottom, left, and right panes. The initial size of the panes can be specified, enhancing the flexibility of the TUI layout. The TUI test script is updated to demonstrate the new TUISplit functionality, ensuring proper integration and usage.

sinkingsugar committed:
Reorder and clean up includes in tui.cpp for improved organization

This commit reorganizes the include statements in tui.cpp by moving the ftxui component includes to follow the shards includes. This change enhances code readability and maintains a consistent structure in the file.

sinkingsugar committed:
Enhance TUIText component with Centered parameter

This commit introduces a new Centered parameter to the TUIText struct, allowing users to center the text within the TUI context. The rendering logic is updated to apply this centering based on the parameter. Additionally, the TUIInput struct is modified to improve input handling with enhanced styling options for better user experience.

sinkingsugar committed:
Add Filler component to TUI for flexible layout management

This commit introduces a new Filler struct to the TUI module, allowing users to add filler elements that help manage spacing within the TUI context. The Filler component is designed to integrate seamlessly with existing TUI elements, enhancing layout flexibility. Additionally, the TUI test scripts are updated to demonstrate the usage of the new Filler component, ensuring proper integration and functionality.

sinkingsugar committed:
Refactor TUIText and introduce new TUI modifiers for enhanced layout control

This commit refactors the TUIText component by removing unused parameters and streamlining the rendering logic. Additionally, it introduces new modifier structs (TUIFlex, TUICentered, TUIAlignedRight, TUIBorder) that allow users to apply flex, centering, right alignment, and border styling to TUI elements. The TUI test script is updated to demonstrate the usage of these new modifiers, ensuring proper integration and functionality within the TUI context.

sinkingsugar committed:
Editing file tui.cpp

sinkingsugar committed:
Editing file tui.cpp

sinkingsugar committed:
Inserting into file tui.cpp

sinkingsugar committed:
Editing file tui.cpp

sinkingsugar committed:
Add TUIFrame and TUIScrollable modifiers for enhanced element wrapping

This commit introduces two new modifier structs: TUIFrame, which wraps input elements in a frame, and TUIScrollable, which allows for vertical or horizontal scrolling of input elements. The TUIScrollable struct includes a parameter to toggle between vertical and horizontal scrolling, enhancing layout flexibility within the TUI context. Additionally, the TUI registration function is updated to include these new modifiers.

sinkingsugar committed:
Add ScrollerBase class to TUI for enhanced scrolling functionality

This commit introduces the ScrollerBase class, which provides a customizable vertical scrolling component for TUI elements. The TUIScrollable modifier is updated to utilize ScrollerBase, simplifying the scrolling implementation. The help text for TUIScrollable is also modified to reflect the change in functionality, enhancing user understanding of the component's capabilities.

sinkingsugar committed:
Enhance ScrollerBase and TUIScrollable with additional parameters for improved functionality

This commit updates the ScrollerBase constructor to accept selected and size parameters, allowing for more customizable scrolling behavior. The TUIScrollable modifier is modified to initialize these parameters, ensuring that the scrolling component can maintain state effectively. This enhancement improves the overall usability and flexibility of the TUI scrolling features.

sinkingsugar committed:
Add TUI test scripts for new functionalities

This commit adds test scripts for the TUI module, specifically including 'tui.shs' and 'tui1.shs' to the build workflows for Linux, macOS, and Windows. The scripts are integrated into the respective CI workflows to ensure they are executed during the build process. Additionally, minor adjustments are made to the 'tui.shs' and 'tui1.shs' files to enhance their functionality and consistency in execution.

sinkingsugar committed:
Comment out TUI test scripts in CI workflow for Linux to prevent execution during builds

sinkingsugar committed:
Rework network kcp MTU sizes and window sizes

guusw commented on shards/modules/tui/tui.cpp:
```diff
@@ -0,0 +1,783 @@
+        _onEnter.activate(context, input, output);
+
+        // Clear buffer for next input
+        _buffer.clear();
```
Can you just clear the value on the shards side. what if we don't want to clear on enter?

sinkingsugar committed:
Refactor TUIInput handling for improved placeholder and value management

This commit modifies the TUIInput structure to streamline the assignment of placeholder text and current value. It ensures that the placeholder is only assigned when the current value is empty, enhancing the clarity of input handling. Additionally, redundant code related to buffer management on the 'on_enter' event has been removed, simplifying the input processing logic.
